src/client/mod.rs

```diff
@@ -25,39 +24,58 @@ pub mod builder {
 }
 
 
-pub struct Client {
-    transport: ImapTransport,
-    state: ClientState,
+pub fn connect(server: &str) -> io::Result<ImapConnectFuture> {
+    let addr = (server, 993).to_socket_addrs()?.next().ok_or_else(|| {
+        io::Error::new(io::ErrorKind::Other, format!("no IP addresses found for {}", server))
+    })?;
+    Ok(ImapConnectFuture::TcpConnecting(TcpStream::connect(&addr), server.to_string()))
 }
 
-impl Client {
-    pub fn connect(server: &str, handle: &Handle) -> ConnectFuture {
-        let addr = format!("{}:993", server);
-        let addr = addr.to_socket_addrs().unwrap().next().unwrap();
-        let stream = TcpStream::connect(&addr, handle);
-        ConnectFuture::TcpConnecting(stream, server.to_string())
-    }
+pub trait ImapClient {
+    type Original;
+    type Transport: ImapTransport;
+    fn into_parts(self) -> (Self::Transport, ClientState);
+    fn rebuild(transport: Self::Transport, state: ClientState) -> Self::Original;
 
-    pub fn call(self, cmd: Command) -> ResponseStream {
-        let Self { transport, mut state } = self;
-        let request_id = state.request_ids.next().unwrap();
+    fn call(self, cmd: Command) -> ResponseStream<Self> where Self: ImapClient + Sized {
+        let (transport, mut state) = self.into_parts();
+        let request_id = state.request_ids.next().unwrap(); // safe: never returns Err
         let (cmd_bytes, next_state) = cmd.into_parts();
         let future = transport.send(Request(request_id.clone(), cmd_bytes));
         ResponseStream::new(future, state, request_id, next_state)
     }
 }
 
-pub struct ResponseStream {
-    future: Option<Send<ImapTransport>>,
-    transport: Option<ImapTransport>,
+pub struct TlsClient {
+    transport: ImapTls,
+    state: ClientState,
+}
+
+impl ImapClient for TlsClient {
```

Comment:
So there is (currently) no `ImapClient` that isn't `TlsClient`? Are you expecting there to be? If not, why the trait?

Reply:
So my goal was to untie `ResponseStream` from the fixed `Client` type. `Client` as is is restricted to exactly `ImapTls`, which seems a bit limiting when the real constraint is just `ImapTransport`. I have two use cases in mind: some kind of mock implementation of `ImapTransport` that can help with testing, and an implementation that can deal with non-TLS TCP streams. Also, the way tokio works really necessitated separating the state from the transport because you have to "give away" the transport for a bit while sending; so, another use case might be if you wanted to implement some kind of alternative client where you have more state to keep around. (I'll add some of this to the commit message.)

The alternative approach is here to just parametrize the `Client` type with `T: ImapTransport`, which I implemented and updated mailsync for, but I found that having to parametrize with `ImapTls` is unergonomic boilerplate for what really seems like the most common case.

Then, there is the issue of naming `Client`; given the above, I was looking for clear names for both of the trait and the default `ImapTls` implementation, and this was the best I could come up with. Better suggestions welcome!

Reply:
If you don't expect there to ever be many implementations of `ImapClient`, only of `ImapTransport`, I'd say the "right" pattern is to have `ImapClient` be a struct parameterized by `T: ImapTransport` (as you outline above). It's true that this puts some burden on users of the library though. There are two workarounds for that:

 - Provide a newtype wrapper. For example:
   
   ```rust
   pub type TlsClient = ImapClient<TlsTransport>;
   ```
 - Store the transport as a `Box<ImapTransport>` inside the `ImapClient`. That way, `ImapClient` does not need to be generic, and the (marginal) cost of dynamic dispatch for network calls.

Reply:
Why would it be better, in your opinion, to have `ImapClient` be a parameterized struct over a trait? This at least doesn't support the third use case I mentioned, and I'm not seeing that many downsides from the current arrangement.

Reply:
I guess to me the biggest downside of a trait here is that I now can't have, say, a struct that holds an `ImapClient`. Instead, it'll either have to hold a `TlsClient`, a `Box<ImapClient>`, or be generic over `T: ImapClient`, all of which seem a little weird. Instead, holding an `ImapClient` (or potentially an `ImapClient<T>`, though I don't think that's necessary) feels more natural somehow.

Reply:
Yeah, and so I'd argue, given what I said above, that that is on purpose, and this design should point you in the direction of instead `impl ImapClient for YourClient` -- but maybe that is too onerous/unidiomatic/weird. Conceptually, given my experience with the tokio stack, it feels like the right design. Can you comment on how that would fit your use cases?

Reply:
(Actually, that might not fit with what I have now -- before, `ClientState` was abstracted out as an associated type `ImapClient::State`, which I think would make this more interesting. Would need some API to get a `&mut ClientState`, I guess.)

Reply:
I'm confused -- surely it would be very uncommon for users of `tokio-imap` to ever `impl ImapClient` for their own things anywhere? They're likely to write programs that *use* an IMAP connection, probably in a binary, but not likely to write their own `ImapClient`?

I think of traits as behaviors. `ImapClient` as a trait means "things that can behave as an `ImapClient`. It's true that `TlsImapClient` can behave as an `ImapClient`, but it does so in exactly the same way that a `MockImapClient` behaves like one (i.e., none of them overwrite `call`). The *real* difference is just in their transport. Because of that I'd argue `ImapClient<T: ImapTransport>` makes more sense. Now, you have a single, shared implementation of an IMAP client that can operate on top of anything that can take IMAP commands and return IMAP responses.

As I argued above, you could then even get rid of the `<T>` by having `ImapClient` store a `Box<ImapTransport>` internally.

Reply:
I'm not sure how `ClientState` factors into this. Do you expect users of the library to ever fiddle with `ClientState` directly?

Reply:
@jonhoo so note that `ImapClient::call()` takes ownership of self (and goes on to destroy it). To me, that makes just putting `TlsClient` in your structs pretty painful. Best case, you have to wrap it in an `Option` and work with `take()` to take it out and put it in again. `impl ImapClient` is one way to make this more explicit and maybe slightly less painful.

Users of the library might want to inspect the `ClientState.state`, but otherwise they don't have to fiddle with it. However, `ImapClient::call()` and `ResponseStream` need access to it.

Reply:
Ah, yes, the terrible terrible "move out of `self`" part of futures. Yeah, I guess maybe you're right that having the user write `impl ImapClient` would make that easier. It's just really weird to me, because their wrapper type that *uses* and `ImapClient` isn't itself an `ImapClient`. Not really. I think I'd actually argue that the right thing to do is to let the user *choose* how they want to use the `ImapClient` you give them, with the knowledge that it consumes and returns `self`.

The internals of `ImapClient::call` and `ResponseStream` are both hidden from the users, so it's fine for them to touch `ClientState`. I was more trying to understand your point about `ClientState` in https://github.com/djc/tokio-imap/pull/14#discussion_r167351289.

Reply:
I tend to favor Jon’s alternative `ImapClient<T>` in principle/i.e. because of the intuitiveness of it. Dirkjan, you explain the `call` method consumes `self`, but does it need to? Can’t it e.g. call `rebuild` and return both the `ResponseStream` and e.g. `self.transport`?

More generally, we can simply stick with Dirkjans current design instead of ruminating on it, since Tokio is in flux anyway and `tokio-io` will be history.

Reply:
`ResponseStream` consumes the transport while it is in the process of sending the request. This could be worked around by adding a buffer, which might improve the situation a bit but will add more errors to be thrown if the buffer is full. In this sense the current API makes it statically impossible to send while no buffer is available, which I like.